pkg/controller/ec2/vpcendpoint/setup.go

```diff
@@ -0,0 +1,380 @@
+package vpcendpoint
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	svcsdk "github.com/aws/aws-sdk-go/service/ec2"
+	svcsdkapi "github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	cpresource "github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/ec2/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+// SetupVPCEndpoint adds a controller that reconciles VPCEndpoint.
+func SetupVPCEndpoint(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(svcapitypes.VPCEndpointGroupKind)
+	opts := []option{setupExternal}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&svcapitypes.VPCEndpoint{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.VPCEndpointGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+func setupExternal(e *external) {
+	c := &custom{client: e.client, kube: e.kube}
+	e.delete = c.delete
+	e.preCreate = preCreate
+	e.postCreate = postCreate
+	e.postObserve = postObserve
+	e.isUpToDate = isUpToDate
+	e.preUpdate = c.preUpdate
+	e.postUpdate = postUpdate
+	e.filterList = filterList
+}
+
+type custom struct {
+	kube   client.Client
+	client svcsdkapi.EC2API
+}
+
+func preCreate(_ context.Context, cr *svcapitypes.VPCEndpoint, obj *svcsdk.CreateVpcEndpointInput) error {
+	// set external name as tag on the vpc endpoint
+	resType := "vpc-endpoint"
+	key := "Name"
+	value := meta.GetExternalName(cr)
+
+	// Clear SGs, RTs, and Subnets if they're empty
+	if len(cr.Spec.ForProvider.SecurityGroupIDs) == 0 {
+		obj.SecurityGroupIds = nil
+	}
+	if len(cr.Spec.ForProvider.RouteTableIDs) == 0 {
+		obj.RouteTableIds = nil
+	}
+	if len(cr.Spec.ForProvider.SubnetIDs) == 0 {
+		obj.SubnetIds = nil
+	}
+
+	// Set tags
+	spec := svcsdk.TagSpecification{
```

Comment:
this should be a loop like this?
```go
import	"github.com/crossplane/crossplane-runtime/pkg/resource"

	for k, v := range resource.GetExternalTags(mg) {
```

it is typically done via `SetupFoo`:
```
			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient()), &tagger{kube: mgr.GetClient()}),
```


https://github.com/crossplane/provider-aws/blob/3f1b7187455f14c306b3037312a19a791ac9d9eb/pkg/controller/ec2/address/controller.go#L222-L248

but I have to say that is a lot of duplicate code, so maybe we should discuss a generic `type Tagger interface { SetTag(k, v string) GetTag(k string) }` with @muvaf etc.

Reply:
This is resolved. Tagging is moved to using an Initializer.